controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/DeploymentQueue.java

```diff
@@ -0,0 +1,157 @@
+     * @param retry Whether this is a retry, and should have increased priority
+     */
+    public void addJob(ApplicationId applicationId, JobType jobType, boolean force, boolean applicationVersionUpgrade, boolean retry) {
+        locked(jobType, queue -> {
+            if (queue.stream().noneMatch(triggering -> triggering.applicationId().equals(applicationId)))
+                queue.add(new Triggering(applicationId, controller.clock().instant(), force, applicationVersionUpgrade, retry));
+        });
+    }
+
+    /** List all jobs currently enqueued. */
+    public List<BuildJob> jobs() {
+        ImmutableList.Builder<BuildJob> builder = ImmutableList.builder();
+        for (JobType jobType : JobType.values())
+            for (Triggering triggering : curator.readJobQueue(jobType))
+                toBuildJob(triggering.applicationId(), jobType).ifPresent(builder::add);
+
+        return builder.build();
+    }
+
+    /** Remove and return a set of jobs to run. This set will contain only one of each job type for capacity constrained zones. */
+    public List<BuildJob> takeJobsToRun() {
+        ImmutableList.Builder<BuildJob> builder = ImmutableList.builder();
+        for (JobType jobType : JobType.values())
+            locked(jobType, queue ->
+                    queue.stream()
+                            .sorted()
+                            .limit(isCapacityConstrained(jobType) ? 1 : Long.MAX_VALUE)
+                            .peek(triggering -> toBuildJob(triggering.applicationId(), jobType).ifPresent(builder::add))
+                            .forEach(queue::remove));
+
+        return builder.build();
+    }
+
+    /** Remove all enqueued jobs for the given application. */
+    public void removeJobs(ApplicationId applicationId) {
+        for (JobType jobType : JobType.values())
+            locked(jobType, queue -> {
+                queue.removeIf(triggering -> triggering.applicationId().equals(applicationId));
+            });
+    }
+
+    /** Lock the job queues and read, modify, and store the queue for the given job type. */
+    private void locked(JobType jobType, Consumer<Collection<Triggering>> modifications) {
+        try (Lock lock = curator.lockJobQueues()) {
+            List<Triggering> queue = new CopyOnWriteArrayList<>(curator.readJobQueue(jobType));
+            modifications.accept(queue);
+            curator.writeJobQueue(jobType, queue);
+        }
+    }
+
+    private static boolean isCapacityConstrained(JobType jobType) {
+        return jobType == JobType.stagingTest || jobType == JobType.systemTest;
+    }
+
+    private Optional<BuildJob> toBuildJob(ApplicationId applicationId, JobType jobType) {
+        return controller.applications().get(applicationId)
+                .flatMap(application -> application.deploymentJobs().projectId())
+                .map(projectId -> new BuildJob(projectId, jobType.jobName()));
+    }
+
+
+    static class Triggering implements Comparable<Triggering> {
```

Comment:
Name doesn't really describe what this is. `QueuedJob` maybe? Or just `Item` since this is an inner class, the fully qualified name would read `DeploymentQueue.Item`.

I suggest making `BuildService.BuildJob` an interface with methods `projectId` and `jobName` and let this implement that interface. If you also let `addJob` take an `Application` instead of `ApplicationId`, you get `projectId` for free. This has several benefits:

* `toBuildJob` disappears and you don't have to do expensive `controller.applications().get(applicationId)` calls for every item that is read from the queue. Dependencies can then be reduced, `controller` field can be replaced a clock.
* The decision to ignore applications without `projectId` can be taken in `addJob`
* Less code, one class instead of two representing the same thing. While keeping visibility of fields only used to sort the queue inside this class.

Reply:
Yes, I already came to the conclusion I should merge those two value classes, as well. 

Getting the `projectId` when getting an item from the queue ensures the `Application` still exists -- we already have applications where the `projectId` is wrong, and this is OK, but I'm not sure what would happen if we were to trigger deployment of an application which was deleted in the meantime. Most likely this is stopped in deployment, and thus isn't a problem. I'll check that, and if it's OK, I'll be happy to simplify.